dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp

```diff
@@ -362,6 +350,28 @@ void PageDirectoryFactory<Trait>::applyRecord(
     }
 }
 
+template <typename Trait>
+void PageDirectoryFactory<Trait>::updateMaxIdByRecord(
+    const PageDirectoryPtr & dir,
+    const typename PageEntriesEdit::EditRecord & r)
+{
+    if constexpr (std::is_same_v<Trait, universal::FactoryTrait>)
+    {
+        // We only need page id under specific prefix after restart.
+        // If you want to add other prefix here, make sure the page id allocation space is still enough after adding it.
+        if (UniversalPageIdFormat::isType(r.page_id, StorageType::Data)
+            || UniversalPageIdFormat::isType(r.page_id, StorageType::Log)
+            || UniversalPageIdFormat::isType(r.page_id, StorageType::Meta))
+        {
+            dir->max_page_id = std::max(dir->max_page_id, Trait::PageIdTrait::getU64ID(r.page_id));
+        }
+    }
+    else
+    {
+        dir->max_page_id = std::max(dir->max_page_id, Trait::PageIdTrait::getU64ID(r.page_id));
+    }
+}
```

Comment:
move to anonymous namespace?

Reply:
This function depends on the `typename Trait` of PageDirectoryFactory. If we make it a static function outside the `PageDirectoryFactory` class, the function signature is complicated.

Reply:
get it